Add viscoelastic material model and elastic outputs functionality

naliboff opened the pull request:
The next step is to incorporate the right-hand side terms.

tjhei commented on include/aspect/elastic.h:
```diff
@@ -0,0 +1,72 @@
+    {
+      public:
+        ElasticOutputs (const unsigned int n_points,
+                     const unsigned int /*n_comp*/)
```
remove ``n_comp``

tjhei commented on include/aspect/elastic.h:
```diff
@@ -0,0 +1,72 @@
+                     const unsigned int /*n_comp*/)
+        {
+          elastic_viscosities.resize(n_points);
+          elastic_evolutions.resize(n_points);
```
reinit ``elastic_evolutions``

tjhei commented:
do you want to derive from AdditionalMaterialOutputsStokesRHS instead?

naliboff commented:
The viscoelastic code currently matches a time-dependent analytical solution for a few time steps , but eventually the solution blows up. No need for a review until I have a better handle on what exactly is going awry.

naliboff commented:
With commit https://github.com/geodynamics/aspect/pull/1593/commits/264331259ebd2c8317710a32714f39e8c42e5733 the code is now passing a time-dependent analytical benchmark. Still some more work to do before a formal review (additional tests, particles), but I think the code structure is now largely correct if anyone would like to try it with a specific test case.

naliboff commented:
Update and question:
1. The current implementation only matches analytical solutions for problems with a fixed strain-rate. I think I know what the issue is, but the fix will require accessing basis function values (`grad_phi_u`) within the material model.
2. Is there any major pitfall or impediment to accessing `grad_phi_u` through `finite_element_values` as in `stokes.cc`, or is there already a way to access these values through the material model interface?

bangerth commented:
The `MaterialModelInputs` structure, as far as I know, only gives you access to the strain rate (i.e., the symmetric gradients). Do you really need the real gradient?

naliboff commented:
I'm pretty sure what other people have used is the real gradient, but I will start with the symmetric gradients to get the new code structure in place and see how this affects the results.

bangerth commented:
Hm, interesting. The reason why the strain is used in most material laws (and in the Stokes and the elasticity equations) is that it is a tensor that rotates appropriately with the coordinate system. In other words, quantities such as the scalar strain rate are independent of the coordinate system. This is not the case for the gradient of the velocity -- it depends on the coordinate system, and consequently is not usually used in material descriptions.

naliboff commented:
@bangerth Everything you outlined above makes sense. For now I've reverted use the symmetric gradient velocity grades (`grads_phi_u`) as this is what was used in other deal.II elasticity examples and I don't have intuitive sense yet for why one would use the full gradient. I'll revisit this after a discussion with @cedrict.

I've added a number of changes and can now reproduce additional benchmarks (ex: viscoelastic bending beam, ..). However, still some work to do condensing commits, adding additional features and  improving the benchmark cases. I actually removed a few benchmark cases that will be added in with separate pull requests.

For now, is it a reasonable strategy to keep this pull request as simple as possible and then continue adding complexity with additional pull requests? 

The large section of code that remains to be added is an option to track the elastic stresses with tracers.

naliboff commented:
This should be ready for a full review now. There have been a significant number of changes since the last pull request and I condensed everything down to 2 commits to help clean things up. I'll add in two benchmarks with a separate commit.

jdannberg commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -231,6 +232,7 @@ namespace aspect
                 data.local_rhs(i) += (force->rhs_u[q] * scratch.phi_u[i]
+                                      + double_contract(force->rhs_e[q],Tensor<2,dim>(scratch.grads_phi_u[i]))
                                       + pressure_scaling * force->rhs_p[q] * scratch.phi_p[i])
                                      * JxW;
```
This means you will always assemble both the force vectors and the elasticity terms, even if only one of them is switched on (e.g. by setting `enable_elasticity` to true). At the moment, this works because the vectors are initialized with zeroes. But it means that, for example, if someone uses your new material model and also the force vector right-hand side, they will always have elasticity, if they want it or not (because you always fill the terms in the material model, no matter what the value of `enable_elasticity` is). So I don't think is is how it's intended to work. 

Maybe it would be better to have two if statements, one assembling the force terms 
`if (force != NULL && this->get_parameters().enable_additional_stokes_rhs)`
and one assembling the elasticity terms
`if (force != NULL && this->get_parameters().enable_elasticity)`.

jdannberg commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -264,6 +266,18 @@ namespace aspect
+      Assert(!this->get_parameters().enable_elasticity
+             ||
+             outputs.template get_additional_output<MaterialModel::AdditionalMaterialOutputsStokesRHS<dim> >()->rhs_e.size()
+             == n_points, ExcInternalError());
```
You can combine all of your new lines with the ones below for the right-hand side, as you do exactly the same things for elasticity as for additional_stokes_rhs. So no need to copy all of that, just combine the if-statements and the Asserts.

jdannberg commented on tests/viscoelastic_fixed_elastic_time_step.prm:
```diff
@@ -0,0 +1,132 @@
+#  Global parameters
```
Can you add a short description of what the test is supposed to do?

jdannberg commented on tests/viscoelastic_fixed_elastic_time_step.prm:
```diff
@@ -0,0 +1,132 @@
+set End time                               = 1e3
+set Use years in output instead of seconds = true
+set Linear solver tolerance                = 1e-7
+set Nonlinear solver scheme                = IMPES
```
Why do you use IMPES for your test cases, isn't the problem nonlinear?

naliboff replied:
In this case I believe the problem is linear as there is no plasticity or non-linear viscous flow terms. Would one need to use the iterative solver simply due to having the stress (strain-rate dependent) in the constitutive law?

jdannberg replied:
I guess I never thought about that, as models with elasticity, but without plasticity don't come up very often. But I think as soon as your viscosity depends on the solution variables of the Stokes system, the problem is nonlinear. 
You can easily test this by just using the iterated IMPES solver, and if the nonlinear solver converges after the second iteration with a very small residual, the problem is linear; otherwise you probably need the nonlinear solver.

jdannberg commented on tests/viscoelastic_numerical_elastic_time_step.prm:
```diff
@@ -0,0 +1,132 @@
+#  Global parameters
```
Can you add a description here too?
Something like, this test is exactly like the test XXX, with the only difference being...

I also wondered: The end time of the model is exactly the same as your elastic time step, so both tests do exactly the same thing. Or is this what you're testing? Wouldn't it make sense to have a model where the elastic time step is different from the real time step, and then see if the stress averaging option works and accounts for the differences?

jdannberg replied:
I also think it would make sense to add the statistics files to the test outputs, because they include much more information (and digits), so it's much easier to see when things have changed and something is different than before.

jdannberg commented on tests/viscoelastic_stress_averaging.prm:
```diff
@@ -0,0 +1,132 @@
+#  Global parameters
```
Same here.

naliboff commented:
@jdannberg - Thanks for the review! I'll likely go through most of the changes tomorrow, but a few initial replies are forthcoming.

naliboff commented on include/aspect/material_model/viscoelastic.h:
```diff
@@ -0,0 +1,232 @@
+     *
+     * The first 3 (2D) or 6 (3D) compositional fields are used to track
+     * individual components of the viscoelastic stress tensor, while additional
+     * fields represent distinct rock types.
```
I agree the naming convention is much safer and I actually meant to do this before updating the pull request. The only advantage to the current method is that it only requires two lines of codes, but it certainly could lead to all sorts of issues. However, the user will still have to be quite careful when assigning properties to a large number of fields in 3D (>10 in some cases?). Would it be worth requiring both a specific naming scheme and ordering as an extra 'safety' feature?

gassmoeller replied:
For now I am fine either way. I can imagine another plugin requiring that its special properties have to be fields 0..2, then we have a problem. But that could be resolved when it finally crops up. Make sure you require the existence of the names though, otherwise allowing for variable ordering but requiring the names in the future will break compatibility.

naliboff commented on source/material_model/viscoelastic.cc:
```diff
@@ -0,0 +1,507 @@
+          const double esm_avg = average_value ( volume_fractions, elastic_shear_moduli, viscosity_averaging);
+
+          //Compute viscoelastic (e.g., effective) viscosity (equation 28 in Moresi et al., 2003, J. Comp. Phys.)
+          const double vev_avg = ( vis_avg * dte ) / ( dte + ( vis_avg / esm_avg ) );
```
Hmm, I honestly had not considered that someone might use this material model with elasticity. I think your suggestion for returning the average of the regular viscosities makes sense if elasticity is not enabled.

I also agree that it would be best not calculate the effective viscosities more than once. This calculation will need to go in a function regardless, as the particle routines will need to access the effective viscosity value in order to update the stresses.

naliboff commented on source/material_model/viscoelastic.cc:
```diff
@@ -0,0 +1,507 @@
+                           ?
+                           this->get_timestep()
+                           :
+                           fixed_elastic_time_step * year_in_seconds );
```
Yep, this is incorrect. I switched around the conditional statement at some point, but did not catch the error as the benchmarks had maximum time steps equivalent equal elastic time steps.

gassmoeller replied:
If you use the fixed elastic time step is the solution still dependent on the model time step as we discussed a few minutes ago?

naliboff commented on source/material_model/viscoelastic.cc:
```diff
@@ -0,0 +1,507 @@
+
+          // Get current and old (previous time step) velocity gradients
+          const QGauss<dim> quadrature_formula (this->get_fe().base_element(this->introspection().base_elements.velocities).degree+1);
+          FEValues<dim> fe_values (this->get_mapping(),
```
I'm not sure either. I'll switch to `Functions::FEFieldFunction` to keep things consistent.

jdannberg replied:
I am not actually sure that my version from the melt models is the better one. 
@tjhei? @bangerth? @gassmoeller? Any comments on that?

bangerth replied:
If you know the reference cell coordinates of the points at which you want to evaluate the solution (e.g., the quadrature points), then `FEValues` is the most efficient way to do this.

You should only use `FEFieldFunction` if you want to evaluate a finite element function at arbitrary points in the domain. This is specifically the case if you only know these points in real space, in which case the procedure of evaluating the solution involves finding the correct cell, and then on that cell finding what reference coordinates correspond to the point where you want the solution.

Does this help? I don't know what you want to do here, but it looks very much like you want to evaluate the function on quadrature points on a given cell.

jdannberg replied:
We want to evaluate the old solution in whatever the evaluation points are that the material model gets as inputs (but only if the material model also gets the current cell, so we don't have to search for that). But I guess that means we really have to use the `FEFieldFunction` here, as we can not know what point the material model gets as an input, and those points do not necessarily have to be the quadrature points.

jdannberg commented on include/aspect/simulator/assemblers/interface.h:
```diff
@@ -110,6 +110,7 @@ namespace aspect
           std::vector<types::global_dof_index> local_dof_indices;
           std::vector<unsigned int>            dof_component_indices;
           std::vector<SymmetricTensor<2,dim> > grads_phi_u;
+          std::vector<Tensor<2,dim> >          grad_phi_u;
```
Sounds good, you know best which formulation makes more sense. I guess the two things to consider are
(1) Which one is more reasonable physically? (Have you discussed that with Cedric? Would it make sense to write a comment about that somewhere?)
(2) What do other people use? (The way I understand your comment above, right at the start of the pull request discussion, other people use the real gradient instead of the symmetric gradient, so one question would be if we need the real gradient to reproduce benchmarks. And I guess we've learned it the hard way with the BA/ALA/TALA approximations that users don't trust the code unless there's the option to reproduce exactly what other people have done.)

naliboff commented:
@jdannberg, @bangerth  - Thanks for the additional discussion on various points above. @cedrict and I are having a follow-up conversation and comparing results, which I will summarize here when we are satisfied. I'll start pushing various updates in the next few days for the benchmarks, tests and stylistics/re-organizational requests.

naliboff commented on source/material_model/viscoelastic.cc:
```diff
@@ -0,0 +1,507 @@
+              // Fill elastic outputs (see equation 30 in Moresi et al., 2003, J. Comp. Phys.)
+              if (force)
+                {
+                  force->rhs_e[i] = -1. * ( ( vev_avg / ( esm_avg * dte  ) ) * stress_old );
```
I decided that the best option is to enforce setting `enable_elasticity` to true through an exception. If the user would like to not have any elastic effects, they can modify the shear modulus so that the viscosity and viscoelastic viscosity are effectively equal.

naliboff commented on source/material_model/viscoelastic.cc:
```diff
@@ -0,0 +1,507 @@
+                             "for background material and compositional fields, "
+                             "for a total of N+1 values, where N is the number of compositional fields. "
+                             "The default value of 75 GPa is representative of mantle rocks. Units: none.");
+          prm.declare_entry ("Use fixed elastic time step", "false",
```
I think the safest route is to use the `CFL number` and `Maximum time step` to ensure the numerical and elastic time step values are equal. This is done in one of the tests and the `Use fixed elastic time step` description also suggests this as an option. I think a good place to further describe this as an option is in the viscoelastic benchmarks and any related cookbooks.

naliboff commented:
@jdannberg - I think I've addressed most of the comments and I've also added two benchmarks. I need to revisit the `FEFieldFunction` method for obtaining previous solutions, but I'm not sure now whether that is necessary. Something to discuss.

The last major item on the to-do list is the particle functionality, but I think that should go in a separate pull request.

jdannberg commented:
Okay, great, I'll have a look later this week!
(Before, I still have to look at @gassmoeller's open pull requests I promised to review weeks ago...)

gassmoeller commented on source/material_model/viscoelastic.cc:
```diff
@@ -0,0 +1,696 @@
+      if (in.current_cell.state() == IteratorState::valid && this->get_timestep_number() > 0 && in.strain_rate.size() > 0)
+        {
+          // Get old (previous time step) velocity gradients
+          const QGauss<dim> quadrature_formula (this->get_fe().base_element(this->introspection().base_elements.velocities).degree+1);
```
This is dangerous because it assumes the quadrature never changes. But it does, e.g. for the postprocessors you will have positions at the vertices instead of the quadrature points (as for the assembly). The solution would be to use the mapping and cell information to convert the global positions into reference cell coordinates and create a quadrature at those positions (we do not care about the quadrature weights here). You could replace this and the following line by this (ignore indentation):
```
std::vector<Point<dim> > quadrature_positions(in.position.size());
for (unsigned int i=0; i < in.position.size(); ++i)
  quadrature_positions[i] = this->get_mapping().transform_real_to_unit_cell(in.current_cell, in.position[i]);

 FEValues<dim> fe_values (this->get_mapping(),
                                            this->get_fe(),
                                            Quadrature<dim>(quadrature_positions),
                                            update_gradients);
```

This will be somewhat slower, but hopefully not dramatically, but it will be much safer. And actually I suspect that your current output it slightly wrong, and will be more accurate with this change.

gassmoeller commented on benchmarks/viscoelastic_bending_beam/viscoelastic_bending_beam.prm:
```diff
@@ -0,0 +1,181 @@
+# compositional fields. Significantly, the time step is limited to 1 Kyr through
+# the "Maximum time step" parameter and a relatively high (0.5) CFL value. Using
+# a constant time step ensures the effective (viscoelastic) viscosity of the
+# viscoelastic beam and viscous fluid remains constants throughout the model run.
```
Does that mean the solution is time step dependent otherwise? Is there a separate elastic timestep from the convection timestep?

gassmoeller commented on benchmarks/viscoelastic_bending_beam/viscoelastic_bending_beam.prm:
```diff
@@ -0,0 +1,181 @@
+set Output directory                       = output
+set Timing output frequency                = 1
+set Pressure normalization                 = surface
+set Surface pressure                       = 0.
```
Can you remove those parameters from the list that are set to default values and not very interesting for this problem? I am thinking of use direct solver, linear solver tolerance, nonlinear solver scheme,  nonlinear solver tolerance, timing output frequency.

naliboff replied:
:+1:

gassmoeller commented on benchmarks/viscoelastic_bending_beam/viscoelastic_bending_beam.prm:
```diff
@@ -0,0 +1,181 @@
+  set Free surface boundary indicators        = 
+  set Tangential velocity boundary indicators = bottom, top, right
+  set Zero velocity boundary indicators       = left
+  set Enable elasticity                       = true
```
I would like to remove this subsection in #2105. Would you be ok with moving the parameter into the `Formulation` subsection that handles all modifications to the equations?

gassmoeller commented on benchmarks/viscoelastic_bending_beam/viscoelastic_bending_beam.prm:
```diff
@@ -0,0 +1,181 @@
+  set Temperature polynomial degree           = 1
+  set Use locally conservative discretization = false
+  set Use discontinuous temperature discretization = false
+  set Use discontinuous composition discretization = true
```
Oh, so you are using discontinuous elements for the stresses as well. Did that cause any problems? Just curious.

gassmoeller commented on benchmarks/viscoelastic_stress_build-up/viscoelastic_stress_build-up.prm:
```diff
@@ -0,0 +1,186 @@
+# or vertical viscoelastic stress magnitude of ~ 200 MPa after 250 Kyr 
+# of deformation. Significantly, the effective "viscoelastic" viscosity
+# is enforced to be constant through time by using a constant time step.
+# This is achieved by setting a maximum time step (1000 years) much lower 
```
much lower *than*

gassmoeller commented on benchmarks/viscoelastic_stress_build-up/viscoelastic_stress_build-up.prm:
```diff
@@ -0,0 +1,186 @@
+# This leads to oscillations in the stress field near the boundaries, which
+# decay towards a constant value in the model interior. This phenomen highlights
+# one advantage to tracking viscoelastic stress with tracers, which do not require
+# defining compositional boundary conditions. 
```
well the fields do not need BCs either. Have you tried removing the boundary indicators from `Fixed composition boundary indicators`? Then there is no flux of stress across the boundary. I am not sure what that will create in the model though.

naliboff replied:
I haven't tried removing the composition boundary indicators yet, but will give it a shot. However, I assumed initially that having no boundary conditions at all would give odd results.

gassmoeller commented on benchmarks/viscoelastic_stress_build-up/viscoelastic_stress_build-up.prm:
```diff
@@ -0,0 +1,186 @@
+  set Free surface boundary indicators        = 
+  set Tangential velocity boundary indicators = top, left 
+  set Zero velocity boundary indicators       = 
+  set Enable elasticity                       = true
```
same comment as above

gassmoeller commented on include/aspect/material_model/interface.h:
```diff
@@ -785,6 +785,14 @@ namespace aspect
+         * momentum equation (first part of the Stokes equation) in each
+         * quadrature point.
+         */
+        std::vector<Tensor<2,dim> > rhs_e;
```
By putting these in here you tie them very closely to the other `AdditionalStokesRHS` outputs. I could also imagine moving the `ElasticAdditionalOutputs` here, and placing `rhs_e` into `ElasticAdditionalOutputs`. I think this would make the logic in the assembler simpler. Just a suggestion though, if you prefer it this way I am fine with it.

gassmoeller commented on include/aspect/material_model/viscoelastic.h:
```diff
@@ -0,0 +1,273 @@
+
+#include <aspect/material_model/interface.h>
+#include <aspect/simulator_access.h>
+#include <deal.II/base/parameter_handler.h>
```
I think the last include is unnecessary.

gassmoeller commented on include/aspect/material_model/viscoelastic.h:
```diff
@@ -0,0 +1,273 @@
+     * publications. However, a notable difference between this material
+     * model and that of previous work is the use of compositional fields,
+     * rather than tracers, to track and advect the stress tensor. The
+     * material model will be updated when an option to track and advect
```
Instead of writing this, put it into a TODO, as in `TODO: Add option to track and advect stresses with tracers.` Makes it easier to find.

gassmoeller commented on include/aspect/material_model/viscoelastic.h:
```diff
@@ -0,0 +1,232 @@
+     * are given that value. Other lengths of lists are not allowed. For a given
+     * compositional field the material parameters are treated as constant,
+     * except density, which varies linearly with temperature according to the
+     * thermal expansivity.
```
My view on this is to make the description of the class at the end of the .cc file an identical copy of this documentation. It is meant for the same audience.

gassmoeller commented on source/material_model/viscoelastic.cc:
```diff
@@ -0,0 +1,696 @@
+#include <numeric>
+#include <deal.II/base/signaling_nan.h>
+
+//using namespace dealii;
```
Remove this line

gassmoeller commented on source/material_model/viscoelastic.cc:
```diff
@@ -0,0 +1,696 @@
+                                 "compositional field is called stress_xz."));
+          AssertThrow(this->introspection().compositional_index_for_name("stress_yz") == 5,
+                      ExcMessage("Material model Viscoelastic only works if the sixth "
+                                 "compositional field is called stress_yz."));
```
Please move all of these asserts into `parse_parameters` or `initialize`. We do not need to check this every time the model is called.

naliboff commented:
@gassmoeller - Thanks for the review!

jdannberg commented on source/material_model/viscoelastic.cc:
```diff
@@ -0,0 +1,700 @@
+                                                     average_elastic_shear_modulus,
+                                                     dte);
+
+          out.viscosities[i] = average_viscoelastic_viscosity;
```
You can same some space by just assigning `out.viscosities[i] = calculate_average_viscoelastic_viscosity...`

jdannberg commented on source/material_model/viscoelastic.cc:
```diff
@@ -0,0 +1,700 @@
+            {
+              elastic_out->elastic_shear_moduli[i] = average_elastic_shear_modulus;
+            }
+
```
remove empty line

jdannberg commented on source/material_model/viscoelastic.cc:
```diff
@@ -0,0 +1,700 @@
+                                   update_gradients);
+
+          fe_values.reinit (in.current_cell);
+          std::vector<Tensor<2,dim> > old_velocity_gradients (Quadrature<dim>(quadrature_positions).size(), Tensor<2,dim>());
```
use `quadrature_positions.size()` instead of `Quadrature<dim>(quadrature_positions).size()`

jdannberg commented on source/material_model/viscoelastic.cc:
```diff
@@ -0,0 +1,700 @@
+
+          FEValues<dim> fe_values (this->get_mapping(),
+                                   this->get_fe(),
+                                   Quadrature<dim>(quadrature_positions),
```
Can you add a comment that the feValues needs a quadrature, so we give it a quadrature and we use a default quadrature here because we only want to evaluate the solution (or its gradients) and do not need to integrate anything

jdannberg commented on source/material_model/viscoelastic.cc:
```diff
@@ -0,0 +1,700 @@
+              const double average_elastic_shear_modulus =
+                calculate_average_elastic_shear_modulus(composition,
+                                                        elastic_shear_moduli,
+                                                        viscosity_averaging);
```
You can use the one from `elastic_out->elastic_shear_moduli[i]`

naliboff replied:
@jdannberg - For some reason this broke the tests. I am going to leave the current structure for now, but happy to revisit this later.

jdannberg commented on source/material_model/viscoelastic.cc:
```diff
@@ -0,0 +1,700 @@
+              const double dt = this->get_timestep();
+              if (use_fixed_elastic_time_step == true && use_stress_averaging == true)
+                {
+                  stress_new = ( ( 1. - ( dt / dte ) ) * stress_old ) + ( ( dt / dte ) * stress_new ) ;
```
Can you add an Assert in `parse_parameters` that dte is not zero?

jdannberg commented on source/material_model/viscoelastic.cc:
```diff
@@ -0,0 +1,700 @@
+              // Fill reaction terms
+              for (unsigned int j = 0; j < SymmetricTensor<2,dim>::n_independent_components ; ++j)
+                out.reaction_terms[i][j] = -stress_old[SymmetricTensor<2,dim>::unrolled_to_component_indices(j)]
+                                           + stress_new[SymmetricTensor<2,dim>::unrolled_to_component_indices(j)];
```
This will not work with the iterated IMPES scheme. Can you add an Assert in Parse_parameters that makes sure we do not use the material model with iterated IMPES.

naliboff replied:
Done.

jdannberg commented on source/material_model/viscoelastic.cc:
```diff
@@ -0,0 +1,700 @@
+                             "constant value, which can be enforced through the CFL condition and"
+                             "maximum time step parameter. When using a fixed elastic time step that"
+                             "differs the numerical time step it is strongly recommended that the stress"
+                             "averaging scheme is also applied");
```
Dot at the end of sentence

gassmoeller commented on source/material_model/viscoelastic.cc:
```diff
@@ -0,0 +1,700 @@
+      }
+      prm.leave_subsection();
+    }
+
```
usually 3 empty lines between functions

gassmoeller commented on source/material_model/viscoelastic.cc:
```diff
@@ -0,0 +1,700 @@
+    Viscoelastic<dim>::parse_parameters (ParameterHandler &prm)
+    {
+      //not pretty, but we need to get the number of compositional fields before
+      //simulatoraccess has been initialized here...
```
This comment is outdated, and you can remove the subsection below. `this->n_compositional_fields()` should work just fine anywhere you need it.

gassmoeller commented on source/material_model/viscoelastic.cc:
```diff
@@ -0,0 +1,700 @@
+            (new MaterialModel::ElasticAdditionalOutputs<dim> (n_points)));
+        }
+    }
+
```
remove empty line

gassmoeller commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -229,11 +229,15 @@ namespace aspect
                                       + pressure_scaling * force->rhs_p[q] * scratch.phi_p[i])
                                      * JxW;
 
+              if (force != NULL && this->get_parameters().enable_elasticity)
```
If you move `rhs_e` out of `AdditionalMaterialOutputsStokesRHS` and into `ElasticOutputs` then you can create a pointer `elastic_outputs` and replace this condition by `if (elastic_outputs != NULL)`

jdannberg commented on source/material_model/viscoelastic.cc:
```diff
@@ -0,0 +1,700 @@
+                             "maximum time step parameter. When using a fixed elastic time step that"
+                             "differs the numerical time step it is strongly recommended that the stress"
+                             "averaging scheme is also applied");
+          prm.declare_entry ("Fixed elastic time step", "1.e3",
```
Lets think about what exactly this does and then document this parameter a bit more.

jdannberg commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -229,11 +229,15 @@ namespace aspect
                                    * JxW;
 
-              if (force != NULL)
+              if (force != NULL && this->get_parameters().enable_additional_stokes_rhs)
```
If you do what @gassmoeller describes below, you don't need this change anymore.

jdannberg commented on include/aspect/material_model/viscoelastic.h:
```diff
@@ -0,0 +1,280 @@
+          const std::vector<double> &compositional_fields) const;
+
+        /**
+         * Reference temperature for thermal expansion.  All components use
```
Please remove one of the spaces between `expansion.  All`

naliboff commented on include/aspect/material_model/interface.h:
```diff
@@ -775,8 +775,41 @@ namespace aspect
+    {
+      public:
+        ElasticOutputs(const unsigned int n_points)
+          : rhs_e(n_points)
```
@gassmoeller - I'm struggling a bit on this point. I've tried a number a permutations for initiating `rhs_e` (now renamed to `elastic_force`) with signaling NaNs and continue to get compiler errors.

I thought `elastic_force(n_points, numbers::signaling_nan<std::vector<Tensor<2,dim> > >() )` would be the correct structure, but apparently not. Before passing on compiler errors, is the structure even in the ballpark of what you had in mind?

gassmoeller replied:
You are close. I think it should be `elastic_force(n_points, numbers::signaling_nan<Tensor<2,dim> >() )` The first parameter is the number of entries, the second parameter is the value that each entry will be initialized to.

naliboff replied:
Doh. Thanks for the fix (compiles successfully now)! 

The bad news is this leads to a segmentation in the stokes assembler during the first time step, when these values have not been calculated yet. Should I put a flag in the assembler to not due this operation on the first time step? Alternatively, I can simply pass a zero value during the initial time step. 

In some studies people often prescribe an initial stress state, so perhaps this second option is the best way forward in terms of future compatibility?

jdannberg replied:
Yes, I think you should decide what these values should be in the initial time step (probably either zero or the initial stress state) and then fill them accordingly in time step zero. The problem occurs because in line 264 of `viscoelastic.cc`, you only fill the elastic outputs if the time step is larger than 0, so you just have to add a separate if that says how the values should be filled if the time step is zero.

naliboff replied:
:+1:

naliboff commented on include/aspect/material_model/interface.h:
```diff
@@ -775,8 +775,41 @@ namespace aspect
+                              const FullMatrix<double>  &/*projection_matrix*/,
+                              const FullMatrix<double>  &/*expansion_matrix*/)
+        {
+          // TODO: not implemented
```
@gassmoeller - Unfortunately the `AssertThrow(false,ExcNotImplemented())` is now executed every time the model runs. Is there a way around this? If not, my vote is to just simply remove `average`.

jdannberg replied:
This should only be executed if you switch on the averaging in your input file, I think. 
If all your models use averaging, then I think we should probably implement this function (it means that you average the viscosity terms on the left-hand side of the equation, but not on the right-hand side...)

naliboff replied:
@jdannberg - Averaging was actually not enabled in this case or any of the other tests and benchmarks (`set Material averaging = none`). This is in the header file, so perhaps that's why the assert is being thrown even though the function is not called?

jdannberg replied:
Oh, I see!
The `average` function is always called, no matter what the averaging operation is. So you could just implement it in a way like:
```
virtual void average (const MaterialAveraging::AveragingOperation operation,
                      const FullMatrix<double>  &/*projection_matrix*/,
                      const FullMatrix<double>  &/*expansion_matrix*/)
  {
      AssertThrow(operation == none,ExcNotImplemented());
      return;
  }
```

naliboff replied:
Hmm, the compiler is complaining that 'operation' was not declared in scope. The other alternative for now is to put an assert in `viscoelastic.cc` that prevents the user from using material averaging.

jdannberg replied:
Did you remove the `/*` and `*/` in line 798?

naliboff replied:
Double doh, time to call it a day ;)

jdannberg replied:
Have a good weekend :smiley:

naliboff commented:
@jdannberg @gassmoeller - Ok, I think the last commit addresses the last round of comments. After you you have had a chance to look at the changes, I'll rebase and fix the tests.

jdannberg commented on source/material_model/viscoelastic.cc:
```diff
@@ -499,7 +509,7 @@ namespace aspect
                                      "compositional field is called stress_xy."));
             }
-          if ( dim == 3)
+          if (dim == 3)
```
shouldn't that be an else if?

jdannberg commented on source/material_model/viscoelastic.cc:
```diff
@@ -252,12 +255,17 @@ namespace aspect
                                                                         dte);
 
-          // Fill elastic outputs if they exist
+          // Fill elastic additional outputs if they exist
```
As we have ElasticOutputs and ElasticAdditionalOutputs (which is a bit confusing) and the code in the lines below makes it clear which of those is filled, I would change this comment to something like "Fill the material properties that are part of the elastic additional outputs (in this case the shear modulus)".

jdannberg commented on source/material_model/viscoelastic.cc:
```diff
@@ -186,6 +184,11 @@ namespace aspect
              MaterialModel::MaterialModelOutputs<dim> &out) const
     {
 
+      // Elastic outputs interface
```
As we have ElasticOutputs and ElasticAdditionalOutputs (which is a bit confusing) I would change this comment to "Create the structure for the elastic force terms that are needed to compute the right-hand side of the Stokes system".

naliboff commented:
@jdannberg - Thanks for taking another look! I addresses the comments, rebased and fixed the tests. I'll rerun the benchmarks tonight to make sure nothing broke in the last few rounds of revision that the tests did not catch.

naliboff commented:
@jdannberg - I reran the benchmarks and everything should be good to go unless another look is needed.

jdannberg commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -207,11 +210,15 @@ namespace aspect
                                       + pressure_scaling * force->rhs_p[q] * scratch.phi_p[i])
                                      * JxW;
 
+              if (elastic_outputs != NULL )
```
This implementation means that in the assembly, if the material model fills the elastic outputs, they will be assembled (no matter what the `enable_elasticity` parameter is set to). I just want to make sure that this is really what we want (as for the force terms, they are only assembled if this option is selected in the input file, see above). Shouldn't this rather be 
`if (elastic_outputs != NULL && this->get_parameters().enable_elasticity)`

jdannberg replied:
At the moment, the model will just crash if you set `enable_elasticity = false`, and the error message says that you have to switch it on if you want to use the elasticity material model. So even if people don't read the documentation, they will notice. 
This would only be to make sure if people add other material models with elasticity that the elastic terms are only added if `enable_elasticity = true`. 

And I think we agreed that if you want to use the material model without elasticity, you just have to set the elastic moduli to a large value. So no need to do any big changes, just this one line. I think we want to merge this as soon as possible.

naliboff replied:
That's right, there is already an Assert for `Enable elasticity`. Probably a sign that is indeed to time to merge as soon as possible :) I've added a list of future features in a new github issue and will address the remaining points in a little bit.

jdannberg commented on tests/viscoelastic_fixed_elastic_time_step.prm:
```diff
@@ -0,0 +1,159 @@
+# This test checks viscoelastic stress buildup with a
```
If this is mostly a copy of an existing benchmark, it would make sense to just include the benchmark in this prm file instead of copying everything (see for example the [mid_ocean_ridge](https://github.com/geodynamics/aspect/blob/master/tests/mid_ocean_ridge.prm) test).

naliboff replied:
This test does derive from the [viscoelastic stress build-up benchmark](https://github.com/naliboff/aspect/blob/06aa82b90890c03e60e63c80ac947194da534bcf/benchmarks/viscoelastic_stress_build-up/viscoelastic_stress_build-up.prm), but there are a number of differences and the other viscoelastic tests are variants from this test. 

My preference is to have one test parameter file with all of the parameters specified. However, I could derive the other two viscoelastic tests from viscoelastic_fixed_elastic_time_step.prm to clean things up.

jdannberg replied:
I think it's fine to have separate parameter files that specify all of the parameters. My only point was, if there is one parameter file that basically represents the benchmark case (maybe just with a lower resolution and shorter end time), it makes sense to include the benchmark prm file rather than copying it, as we will immediately notice when a pull request breaks that test (while the benchmarks are not run for every pull request), so the benchmark prm file will always be working and up to date.

naliboff replied:
I actually used the benchmark case as a template for the test for exactly that reason, but having it derive directly from the benchmark would make that perfectly clear for anyone else who runs into issues. Definitely worth the change :+1:

jdannberg replied:
Yes, and the other big point is that in case we ever do changes to the structure of the parameter file again, we will notice if we have to update the benchmark parameter file (in the past, it happened that we just updated the test prm file, and forgot about the file in the benchmarks folder, as it wasn't used directly in any test).

naliboff commented:
@jdannberg - My preference is to have  `Use fixed elastic time step = false` by default as this is the option most commonly used in the community.

gassmoeller commented:
Hmm, I feel like the `Use fixed elastic time step` parameter is a controversial point. I just discussed this with @jdannberg and I can understand her point that this makes models time-step dependent and nearly impossible to reproduce. On the other hand if it is the commonly used option in the community it would be nice to allow for it. So my suggestion is: Let us force the user to make this decision. Instead of a `Patterns::Bool()` use a `Patterns::Selection("true|false|unspecified")` with `unspecified` as the default setting. Then parse this in the parse_parameters function and add `AssertThrow(parameter != "unspecified",ExcMessage("... (this parameter has to be specified, because ... ))`. Is that a reasonable workaround?

naliboff commented:
@gassmoeller @jdannberg  - Yes, forcing the user to at least make a decision sounds like a good solution. Hopefully at some point soon we can find a new formulation that handles the time-dependence in a more reasonable matter or simply removes it.

jdannberg commented:
Okay, looks good to me. You can merge yourself once the tester is happy :)
And good job, it's great that we can finally merge this!

naliboff commented:
Yahoo! Thanks to you and Rene for all the reviews, feedback and patience!
